internal/mocks/mock_slow_storage.go

```diff
@@ -26,31 +26,32 @@ func NewMockSlowDataStorage(ds storage.OpenFGADatastore, readTuplesDelay time.Du
 
 func (m *slowDataStorage) Close() {}
 
-func (m *slowDataStorage) Read(ctx context.Context, store string, key *openfgav1.TupleKey) (storage.TupleIterator, error) {
+func (m *slowDataStorage) Read(ctx context.Context, store string, key *openfgav1.TupleKey, queryOpts storage.QueryOptions) (storage.TupleIterator, error) {
 	time.Sleep(m.readTuplesDelay)
-	return m.OpenFGADatastore.Read(ctx, store, key)
+	return m.OpenFGADatastore.Read(ctx, store, key, queryOpts)
 }
 
-func (m *slowDataStorage) ReadPage(ctx context.Context, store string, key *openfgav1.TupleKey, paginationOptions storage.PaginationOptions) ([]*openfgav1.Tuple, []byte, error) {
+func (m *slowDataStorage) ReadPage(ctx context.Context, store string, key *openfgav1.TupleKey, paginationOptions storage.PaginationOptions, queryOpts storage.QueryOptions) ([]*openfgav1.Tuple, []byte, error) {
```

Comment:
Apologies for coming in a bit late to this, but seeing this adds a breaking change to the API, I wonder if there's an opportunity for a potential refactor that guards us better against potential other breaking changes in the future and improves the API usage at the same time. 

I personally find it odd that we're passing 2 types of options as a struct over here, what happens if in the future we find a need for another type of option? 

I believe having just 1 Options interface would be cleaner. Here's an example of what I'm suggesting:

```go
type Option interface {
   Apply(*Options)
}

type Options struct {
   Pagination *PaginationOptions
   Consistency *ConsistencyOptions
   // something else in the future 
}

type PaginationOptions struct {
    Limit  int
    Offset int
}

func (p PaginationOptions) Apply(opts *Options) {
    opts.Pagination = &p
}

type ConsistencyOptions struct {
    ConsistencyLevel string
}

func (c ConsistencyOptions) Apply(opts *Options) {
    opts.Consistency = &c
}

func ReadPage(ctx context.Context, store string, tupleKey *openfgav1.TupleKey, options ...Option) ([]*openfgav1.Tuple, []byte, error) {
    opts := &Options{}

    // Apply each option to the opts struct
    for _, option := range options {
        option.Apply(opts)
    }

    // Use the extracted options as needed
    if opts.Pagination != nil {
        // Handle pagination options
    }

    if opts.Consistency != nil {
        // Handle consistency options
    }

    // Function implementation here

    return nil, nil, nil
}

// The usage
func main() {
    ctx := context.Background()
    store := "exampleStore"
    tupleKey := "exampleKey"
    paginationOpts := PaginationOptions{Limit: 10, Offset: 0}
    consistencyOpts := ConsistencyOptions{ConsistencyLevel: "strong"}

   // if we need to pass in any option
    _, _, err := ReadPage(ctx, store, &tupleKey, paginationOpts, consistencyOpts)

   // if we don't need any option
       _, _, err := ReadPage(ctx, store, &tupleKey)
}
```

Reply:
Thank you for this suggestion.  I agree with the suggestion that it makes it more resilient to changes in the future.

Reply:
I've pushed 3be546cea7ecb0ae337a82ddc27c275776d81803 that moves the APIs I already touched to this new pattern.

There's a couple thoughts I have:

* We now have the consequence of passing all the `Read*` APIs a pagination option even if it wasn't there before, not sure if this is a concern.
* It was initially out of scope for this work, but in the spirit of "in for a penny, in for a pound" do we take this chance to roll a similar change out to all APIs that accepts a `PaginationOption` to allow us to make changes like this easily across the APIs?

Reply:
To make sure it's clear what extending to the other functions looks like. I'm proposing extending to `ReadAuthorizationModels`, `ListStores`, and `ChangelogBackend` alongside having different options per type of `Backend` so the `Options` that's currently in this PR would become `RelationshipTupleReaderOptions` and we would also introduce `ReadAuthorizationModelsOptions`, `ListStoresOptions`, `ChangelogBackendOptions` so that we can accurately reflect the options accepted by each backend.

```patch
diff --git a/pkg/storage/storage.go b/pkg/storage/storage.go
index 2193ff0c..38b649dd 100644
--- a/pkg/storage/storage.go
+++ b/pkg/storage/storage.go
@@ -213,7 +213,7 @@ type AuthorizationModelReadBackend interface {
 	ReadAuthorizationModel(ctx context.Context, store string, id string) (*openfgav1.AuthorizationModel, error)
 
 	// ReadAuthorizationModels reads all models for the supplied store and returns them in descending order of ULID (from newest to oldest).
-	ReadAuthorizationModels(ctx context.Context, store string, options PaginationOptions) ([]*openfgav1.AuthorizationModel, []byte, error)
+	ReadAuthorizationModels(ctx context.Context, store string, options ...Option) ([]*openfgav1.AuthorizationModel, []byte, error)
 
 	// FindLatestAuthorizationModel returns the last model for the store.
 	// If none were ever written, it must return ErrNotFound.
@@ -241,7 +241,7 @@ type StoresBackend interface {
 	CreateStore(ctx context.Context, store *openfgav1.Store) (*openfgav1.Store, error)
 	DeleteStore(ctx context.Context, id string) error
 	GetStore(ctx context.Context, id string) (*openfgav1.Store, error)
-	ListStores(ctx context.Context, paginationOptions PaginationOptions) ([]*openfgav1.Store, []byte, error)
+	ListStores(ctx context.Context, options ...Option) ([]*openfgav1.Store, []byte, error)
 }
 
 // AssertionsBackend is an interface that defines the set of methods for reading and writing assertions.
@@ -267,8 +267,8 @@ type ChangelogBackend interface {
 		ctx context.Context,
 		store,
 		objectType string,
-		paginationOptions PaginationOptions,
 		horizonOffset time.Duration,
+		options ...Option,
 	) ([]*openfgav1.TupleChange, []byte, error)
 }
 

 ```

Reply:
After some discussion, we have decided to go ahead with this change to provide a consistent way of passing options across the methods. The PR has been updated with these changes 🎉

Reply:
Can we compromise on having different types of options per method? E.g. `ReadUsersetTuplesOption`, `ReadChangesOption` etc? I.e. fine-grained options 😄 

> Not all the methods support PaginationOptions for example.

This is my concern too. If in the future we introduce a contrib-style repo with new storage adapters that we don't own, people will be puzzled as to whether they need to support pagination or not.